R/createSets.R

```diff
@@ -0,0 +1,267 @@
+
+  bs <- getBrickMapping("buildingShell.csv") %>%
+    getElement("bs") %>%
+    unique()
+  if (config[["ignoreShell"]]) bs <- head(bs, 1)
+  bs <-  m$addSet(
+    "bs",
+    records = bs,
+    description = "building shell"
+  )
+
+  bsr <- m$addSet(
+    "bsr",
+    records = c(bs$getUELs(), 0),
+    description = "renovated building shell"
+  )
+
+
+  ## heating system ====
+
+  hsMap <- getBrickMapping("heatingSystem.csv", "sectoral")
+
+  hs <- hsMap %>%
+    getElement("hs") %>%
+    unique()
+  hs <-  m$addSet(
+    "hs",
+    records = hs,
+    description = "heating system"
+  )
+
+  hsr <- m$addSet(
+    "hsr",
+    records = c(0, hs$getUELs()),
+    description = "renovated heating system"
+  )
+
+  carrier <- hsMap %>%
+    getElement("carrier") %>%
+    unique()
+  carrier <- m$addSet(
+    "carrier",
+    records = carrier,
+    description = "energy carrier"
+  )
+
+  hsCarrier <- unique(hsMap[, c("hs", "carrier")])
+  hsCarrier <- m$addSet(
+    "hsCarrier",
+    c(hs, carrier),
+    records = hsCarrier,
+    description = "mapping between heating system and energy carrier"
+  )
+
+
+
+  # Independent stock subset ---------------------------------------------------
+
+  reg <- m$addSet(
+    "reg",
+    records = config[["regions"]],
+    description = "region"
+  )
+
+  loc <- getBrickMapping("location.csv") %>%
+    getElement("loc") %>%
+    unique()
+  loc <- m$addSet(
+    "loc",
+    records = loc,
+    description = "location of building (rural, urban)"
+  )
+
+  typ <- getBrickMapping("buildingType.csv") %>%
+    getElement("typ") %>%
+    unique()
+  typ <- m$addSet(
+    "typ",
+    records = typ,
+    description = "type of residential building (SFH, MFH)"
+  )
+
+  inc <- getBrickMapping("incomeQuantile.csv") %>%
+    getElement("inc") %>%
+    unique()
+  inc <- m$addSet(
+    "inc",
+    records = inc,
+    description = "income quantile"
+  )
+
+
+
+  # Other ----------------------------------------------------------------------
+
+
+  ## Boiler ban ====
+
+  # read ban definition from config
+  hsBanConfig <- config[["boilerBan"]] %>%
+    listToDf() %>%
+    guessColnames(m)
+
+  if (!(is.null(hsBanConfig) || identical(hsBanConfig, "NULL"))) {
+
+    # list banned periods
+    hsBanConfig <- hsBanConfig %>%
+      group_by(across(everything())) %>%
+      mutate(ttot = head(ttotNum, 1)) %>%
+      complete(ttot = ttotNum) %>%
+      ungroup() %>%
+      filter(.data[["ttot"]] > .data[["value"]]) %>%
+      select(-"value")
+    hsBan <- expandSets(var, reg, ttot, hs) %>%
+      mutate(across(everything(), as.character)) %>%
+      mutate(ttot = as.numeric(.data[["ttot"]]))
+    hsBan <- hsBan %>%
+      inner_join(hsBanConfig, by = intersect(colnames(hsBan),
+                                             colnames(hsBanConfig))) %>%
+      select("var", "reg", "ttot", "hs")
+  } else {
+    hsBan <- NULL
+  }
+  hsBan <- m$addSet(
+    "hsBan",
+    records = hsBan,
+    domain = c(var, reg, ttot, hs),
```

Comment:
This is a bit inconsistent with the definition of multi-dimensional sets above (i.e. vinExists or hsCarrier), where ```domain = ``` is skipped. I would actually say that this is the cleaner way, but that would affect also the parameter definitions.

Reply:
You are right. I made all arguments explicit.